polygon/bor/bor.go

```
@@ -1558,13 +1559,52 @@ func (c *Bor) CommitStates(
 	chain statefull.ChainContext,
 	syscall consensus.SystemCall,
 	logger log.Logger,
+	fetchEventsWithingTime bool,
 ) error {
 	blockNum := header.Number.Uint64()
 
 	if c.useBridgeReader {
-		events, err := c.bridgeReader.Events(c.execCtx, blockNum)
-		if err != nil {
-			return err
+		var events []*types.Message
+		var err error
+
+		if fetchEventsWithingTime {
+			sprintLength := c.config.CalculateSprintLength(blockNum)
+
+			if blockNum < sprintLength {
+				return nil
+			}
+
+			prevSprintStart := chain.Chain.GetHeaderByNumber(blockNum - sprintLength)
+			stateSyncDelay := c.config.CalculateStateSyncDelay(blockNum)
+
+			timeFrom := time.Unix(int64(prevSprintStart.Time-stateSyncDelay), 0)
+			timeTo := time.Unix(int64(header.Time-stateSyncDelay), 0)
+
+			// Previous sprint was not indore.
+			if !c.config.IsIndore(prevSprintStart.Number.Uint64()) {
+				prevPrevSprintStart := chain.Chain.GetHeaderByNumber(prevSprintStart.Number.Uint64() - sprintLength)
+
+				if prevPrevSprintStart != nil {
+					timeFrom = time.Unix(int64(prevPrevSprintStart.Time), 0)
+				} else {
```

`chain.Chain.GetHeaderByNumber` should never return a `nil` header - that is unexpected behaviour and we should return an err in this case